### PR TITLE
Generate smoosh_generated.h in OBJDIR (fixes #34)

### LINKED ISSUE #34: smoosh.h doesn't get updated before use
when `SmooshResult` in `js/src/frontend/smoosh/src/lib.rs` gets updated, `smoosh.h` is supposed to be updated,
but the build system performs a build for `js/src/frontend/Frontend2.cpp` (that includes `smoosh.h`)
before updating `smoosh.h`

### PR DESCRIPTION
**Pull request by arai-a**
Instead of using `build.rs`, use `GeneratedFile` in `moz.build`, with `/layout/style/RunCbindgen.py`.

There's one change outside of `js/`,
in `build/moz.configure/bindgen.configure`, cbindgen isn't configured for js shell, but now we need it, so removed the `when` parameter there.

### CODE REVIEW

**Comment by arai-a** [comment]
now preparing some changes for taskcluster jobs, and package

**Comment by arai-a** [comment]
2 more changes outside js

### layout/style/RunCbindgen.py 

given SpiderMonkey tarball doesn't contain toolkit/library/rust and depending crates, we cannot use it as crate path.  but for smoosh's case, just using the smoosh's crate path itself works
(for other crates in m-c, just using the same crate path causes version mismatch in dependencies)

So, added `GeneratedFile` flags `--use-same-crate` to tell it to use the same crate path as input

### taskcluster/ci/spidermonkey/kind.yml

haven't tested tho, to use cbindgen in JS shell build on automation, added cbindgen to required toolchain

**Comment by arai-a** [comment]
confirmed building with/without --enable-smoosh works with SpiderMonkey tarball created by `make-source-package.sh`.

for without `--enable-smoosh` case (without jsparagus directory), I've commented out local dependency here
https://github.com/mozilla-spidermonkey/rust-frontend/blob/b4df63529b9165bc5a861aae4a93abf0bd09e9f3/js/src/frontend/smoosh/Cargo.toml#L8-L11
see https://github.com/mozilla-spidermonkey/rust-frontend/issues/44 for more details

**Comment by arai-a** [comment]
this fails on try
https://treeherder.mozilla.org/#/jobs?repo=try&revision=9bc18146e0d99c4d6f986d62081b5f01d7463c01&group_state=expanded&selectedJob=289183096

looks like I need some more code to make mach find cbindgen
cbindgen artifact itself is downloaded and extracted to /builds/worker/fetches

**Comment by arai-a** [comment]
passed by specifying CBINDGEN env in autospider.py

**Comment by arai-a** [comment]
still fails on browser build
https://treeherder.mozilla.org/logviewer.html#/jobs?job_id=289187980&repo=try&lineNumber=1371
somehow, the CBINDGEN env passed to toplevel configure isn't passed to configure in js/src :/

File: build/moz.configure/bindgen.configure
Comment (nbp) [inline build/moz.configure/bindgen.configure:8]: Note: I presume this might need some announcement on the JS mailing list once we merge the rust-frontend project in mozilla-central. As from what I understand, this unconditionally add `cbindgen` as a dependency.

File: layout/style/RunCbindgen.py
Comment (nbp) [inline layout/style/RunCbindgen.py]: I will forward this part to @heycam .

**Review by nbp — APPROVED** [review]

File: layout/style/RunCbindgen.py
Comment (arai-a) [reply layout/style/RunCbindgen.py]: I'll post this file as separate patch for m-c.

Comment (arai-a) [reply layout/style/RunCbindgen.py]: posted in https://bugzilla.mozilla.org/show_bug.cgi?id=1616161
I'll merge this once it gets landed

**Comment by arai-a** [comment]
Removed the change to `layout/style/RunCbindgen.py` (assuming that the change will be landed to m-c).
and also updated vendored crates

**Comment by arai-a** [comment]
so, this should not be merged until https://bugzilla.mozilla.org/show_bug.cgi?id=1616161 gets fixed and merged to rust-frontend

**Pull request merged**
